File: packages/tooling/config-validation/validate-config.mjs
import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import Ajv2020 from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

const schemaPath = path.resolve("packages/contracts/config.schema.json");
const schema = JSON.parse(fs.readFileSync(schemaPath, "utf-8"));

const ajv = new Ajv2020({ allErrors: true, strict: false });
addFormats(ajv);
const validate = ajv.compile(schema);

const schemaProps = Object.keys(schema.properties || {});
const config = Object.fromEntries(
  schemaProps
    .map((name) => [name, process.env[name]])
    .filter(([, value]) => value !== undefined)
);
const valid = validate(config);

if (!valid) {
  console.error("Invalid runtime config:");
  console.error(validate.errors);
  process.exit(1);
}

console.log("Runtime config validation passed.");

File: packages/tooling/config-validation/validate_config.py
import json
import os
from pathlib import Path

from jsonschema import Draft7Validator


schema_path = Path("packages/contracts/config.schema.json")
with schema_path.open("r", encoding="utf-8") as f:
    schema = json.load(f)

validator = Draft7Validator(schema)
errors = sorted(validator.iter_errors(dict(os.environ)), key=lambda e: e.path)

if errors:
    print("Invalid runtime config:")
    for error in errors:
        print(f"- {error.message}")
    raise SystemExit(1)

print("Runtime config validation passed.")
